File: wearapp/src/main/java/com/thewizrd/simpleweather/wearable/DataSyncManager.java
package com.thewizrd.simpleweather.wearable;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.core.util.ObjectsCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.android.gms.tasks.Tasks;
import com.google.android.gms.wearable.Asset;
import com.google.android.gms.wearable.DataMap;
import com.google.android.gms.wearable.Wearable;
import com.google.common.base.Function;
import com.google.common.collect.Collections2;
import com.thewizrd.shared_resources.icons.WeatherIconsProvider;
import com.thewizrd.shared_resources.locationdata.LocationData;
import com.thewizrd.shared_resources.tasks.AsyncTask;
import com.thewizrd.shared_resources.utils.CommonActions;
import com.thewizrd.shared_resources.utils.JSONParser;
import com.thewizrd.shared_resources.utils.LocaleUtils;
import com.thewizrd.shared_resources.utils.Logger;
import com.thewizrd.shared_resources.utils.Settings;
import com.thewizrd.shared_resources.utils.StringUtils;
import com.thewizrd.shared_resources.utils.Units;
import com.thewizrd.shared_resources.wearable.WearableHelper;
import com.thewizrd.shared_resources.wearable.WearableSettings;
import com.thewizrd.shared_resources.weatherdata.Forecasts;
import com.thewizrd.shared_resources.weatherdata.HourlyForecast;
import com.thewizrd.shared_resources.weatherdata.HourlyForecasts;
import com.thewizrd.shared_resources.weatherdata.Weather;
import com.thewizrd.shared_resources.weatherdata.WeatherManager;
import com.thewizrd.simpleweather.services.WidgetUpdaterWorker;

import org.checkerframework.checker.nullness.compatqual.NullableDecl;

import java.io.IOException;
import java.io.InputStream;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;

import timber.log.Timber;

public class DataSyncManager {
    @WorkerThread
    public static void updateSettings(@NonNull Context context, final DataMap dataMap) {
        context = context.getApplicationContext();

        if (dataMap != null && !dataMap.isEmpty()) {
            LocalBroadcastManager localBroadcastMgr = LocalBroadcastManager.getInstance(context);

            long updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME);

            if (updateTimeMillis != getSettingsUpdateTime(context)) {
                String API = dataMap.getString(WearableSettings.KEY_API, "");
                String API_KEY = dataMap.getString(WearableSettings.KEY_APIKEY, "");
                boolean keyVerified = dataMap.getBoolean(WearableSettings.KEY_APIKEY_VERIFIED, false);
                if (!StringUtils.isNullOrWhitespace(API)) {
                    Settings.setAPI(API);
                    if (WeatherManager.isKeyRequired(API)) {
                        Settings.setAPIKEY(API_KEY);
                        Settings.setKeyVerified(false);
                    } else {
                        Settings.setAPIKEY("");
                        Settings.setKeyVerified(false);
                    }
                }

                Settings.setFollowGPS(dataMap.getBoolean(WearableSettings.KEY_FOLLOWGPS, false));

                DataMap unitMap = dataMap.getDataMap(WearableSettings.KEY_UNITS);
                final String oldUnits = Settings.getUnitString();
                if (unitMap != null) {
                    Settings.setTemperatureUnit(unitMap.getString(WearableSettings.KEY_TEMPUNIT, Units.FAHRENHEIT));
                    Settings.setSpeedUnit(unitMap.getString(WearableSettings.KEY_SPEEDUNIT, Units.MILES_PER_HOUR));
                    Settings.setDistanceUnit(unitMap.getString(WearableSettings.KEY_DISTANCEUNIT, Units.MILES));
                    Settings.setPressureUnit(unitMap.getString(WearableSettings.KEY_PRESSUREUNIT, Units.INHG));
                    Settings.setPrecipitationUnit(unitMap.getString(WearableSettings.KEY_PRECIPITATIONUNIT, Units.INCHES));
                } else {
                    Settings.setDefaultUnits(dataMap.getString(WearableSettings.KEY_TEMPUNIT, Units.FAHRENHEIT));
                }
                final String newUnits = Settings.getUnitString();

                if (!ObjectsCompat.equals(oldUnits, newUnits)) {
                    localBroadcastMgr.sendBroadcast(new Intent(CommonActions.ACTION_SETTINGS_UPDATEUNIT));
                }

                LocaleUtils.setLocaleCode(dataMap.getString(WearableSettings.KEY_LANGUAGE, ""));

                final String oldIcons = Settings.getIconsProvider();
                Settings.setIconsProvider(dataMap.getString(WearableSettings.KEY_ICONPROVIDER, WeatherIconsProvider.KEY));
                final String newIcons = Settings.getIconsProvider();
                if (!ObjectsCompat.equals(oldIcons, newIcons)) {
                    // Update tiles and complications
                    WidgetUpdaterWorker.requestWidgetUpdate(context);
                }

                setSettingsUpdateTime(context, updateTimeMillis);

                Timber.tag("DataSyncManager").d("Updated settings");
            }

            // Send callback to receiver
            localBroadcastMgr.sendBroadcast(new Intent(WearableHelper.SettingsPath));
        }
    }

    @WorkerThread
    public static void updateLocation(@NonNull Context context, final DataMap dataMap) {
        context = context.getApplicationContext();

        if (dataMap != null && !dataMap.isEmpty()) {
            String locationJSON = dataMap.getString(WearableSettings.KEY_LOCATIONDATA, "");
            if (!StringUtils.isNullOrWhitespace(locationJSON)) {
                LocationData locationData = JSONParser.deserializer(locationJSON, LocationData.class);

                if (locationData != null) {
                    long updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME);

                    if (updateTimeMillis != getLocationDataUpdateTime(context) ||
                            !locationData.equals(Settings.getHomeData())) {
                        Settings.saveHomeData(locationData);
                    }

                    setLocationDataUpdateTime(context, updateTimeMillis);

                    Timber.tag("DataSyncManager").d("updateLocation: Updated location data");

                    // Send callback to receiver
                    LocalBroadcastManager.getInstance(context).sendBroadcast(
                            new Intent(WearableHelper.LocationPath));
                }
            }
        }
    }

    @WorkerThread
    public static void updateWeather(@NonNull Context context, final DataMap dataMap) {
        context = context.getApplicationContext();

        if (dataMap != null && !dataMap.isEmpty()) {
            long updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME);
            // Check if data actually exists to force an update
            boolean dataExists = false;
            LocationData homeData = Settings.getHomeData();
            if (homeData != null) {
                dataExists = Settings.getWeatherDAO().getWeatherDataCountByKey(homeData.getQuery()) > 0;
            }

            if (updateTimeMillis != getWeatherUpdateTime(context) || !dataExists) {
                final Asset weatherAsset = dataMap.getAsset(WearableSettings.KEY_WEATHERDATA);
                if (weatherAsset != null) {
                    try (InputStream inputStream = Tasks.await(Wearable.getDataClient(context).getFdForAsset(weatherAsset)).getInputStream()) {
                        final Weather weatherData = AsyncTask.await(new Callable<Weather>() {
                            @Override
                            public Weather call() {
                                return JSONParser.deserializer(inputStream, Weather.class);
                            }
                        });

                        if (weatherData != null && weatherData.isValid()) {
                            Settings.saveWeatherAlerts(homeData, weatherData.getWeatherAlerts());
                            Settings.saveWeatherData(weatherData);
                            Settings.saveWeatherForecasts(new Forecasts(weatherData.getQuery(), weatherData.getForecast(), weatherData.getTxtForecast()));
                            Settings.saveWeatherForecasts(weatherData.getQuery(), weatherData.getHrForecast() == null ? null :
                                    Collections2.transform(weatherData.getHrForecast(), new Function<HourlyForecast, HourlyForecasts>() {
                                        @NonNull
                                        @Override
                                        public HourlyForecasts apply(@NullableDecl HourlyForecast input) {
                                            return new HourlyForecasts(weatherData.getQuery(), input);
                                        }
                                    }));
                            Settings.setUpdateTime(LocalDateTime.ofInstant(Instant.ofEpochMilli(updateTimeMillis), ZoneOffset.UTC));
                            setWeatherUpdateTime(context, updateTimeMillis);

                            Timber.tag("DataSyncManager").d("Updated weather data");

                            WidgetUpdaterWorker.requestWidgetUpdate(context);
                        } else {
                            Timber.tag("DataSyncManager").d("Weather data invalid");
                        }
                    } catch (ExecutionException | InterruptedException | IOException e) {
                        Logger.writeLine(Log.ERROR, e);
                    }
                } else {
                    Timber.tag("DataSyncManager").d("updateWeather: weather data missing");
                }
            }

            // Send callback to receiver
            LocalBroadcastManager.getInstance(context).sendBroadcast(
                    new Intent(WearableHelper.WeatherPath));
        }
    }

    private static long getSettingsUpdateTime(@NonNull Context context) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        return prefs.getLong("settings_updatetime", 0);
    }

    private static long getLocationDataUpdateTime(@NonNull Context context) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        return prefs.getLong("location_updatetime", 0);
    }

    private static long getWeatherUpdateTime(@NonNull Context context) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        return prefs.getLong("weather_updatetime", 0);
    }

    private static void setSettingsUpdateTime(@NonNull Context context, long value) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        prefs.edit().putLong("settings_updatetime", value).apply();
    }

    private static void setLocationDataUpdateTime(@NonNull Context context, long value) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        prefs.edit().putLong("location_updatetime", value).apply();
    }

    private static void setWeatherUpdateTime(@NonNull Context context, long value) {
        context = context.getApplicationContext();

        SharedPreferences prefs = context.getSharedPreferences("datasync", Context.MODE_PRIVATE);
        prefs.edit().putLong("weather_updatetime", value).apply();
    }
}

File: wearapp/src/main/java/com/thewizrd/simpleweather/wearable/DataSyncManager.kt
package com.thewizrd.simpleweather.wearable

import android.content.Context
import android.content.Intent
import androidx.core.content.edit
import androidx.core.util.ObjectsCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.wearable.DataMap
import com.google.android.gms.wearable.Wearable
import com.google.android.gms.wearable.WearableStatusCodes
import com.thewizrd.common.wearable.WearableHelper
import com.thewizrd.common.wearable.WearableSettings
import com.thewizrd.shared_resources.database.WeatherDatabase
import com.thewizrd.shared_resources.icons.WeatherIconsEFProvider
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.preferences.DevSettingsEnabler
import com.thewizrd.shared_resources.preferences.SettingsManager
import com.thewizrd.shared_resources.utils.CommonActions
import com.thewizrd.shared_resources.utils.JSONParser
import com.thewizrd.shared_resources.utils.LocaleUtils
import com.thewizrd.shared_resources.utils.Units
import com.thewizrd.shared_resources.weatherdata.model.Forecasts
import com.thewizrd.shared_resources.weatherdata.model.HourlyForecasts
import com.thewizrd.shared_resources.weatherdata.model.Weather
import com.thewizrd.simpleweather.services.WidgetUpdaterWorker.Companion.requestWidgetUpdate
import com.thewizrd.weather_api.weatherModule
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.isActive
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import timber.log.Timber
import java.io.IOException
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset

object DataSyncManager {
    private const val TAG = "DataSyncManager"

    suspend fun updateSettings(context: Context, dataMap: DataMap?) = withContext(Dispatchers.Default) {
        val appContext = context.applicationContext
        val settingsMgr = SettingsManager(appContext)

        if (dataMap != null && !dataMap.isEmpty) {
            val localBroadcastMgr = LocalBroadcastManager.getInstance(appContext)

            val updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME)

            if (!isActive) return@withContext

            if (updateTimeMillis != getSettingsUpdateTime(appContext)) {
                val API = dataMap.getString(WearableSettings.KEY_API, "")
                val API_KEY = dataMap.getString(WearableSettings.KEY_APIKEY, "")
                val keyVerified = dataMap.getBoolean(WearableSettings.KEY_APIKEY_VERIFIED, false)
                if (!API.isNullOrBlank()) {
                    settingsMgr.setAPI(API)
                    if (weatherModule.weatherManager.isKeyRequired(API)) {
                        settingsMgr.setAPIKey(API, API_KEY)
                        settingsMgr.setKeyVerified(API, keyVerified)
                    } else {
                        settingsMgr.setAPIKey(API, "")
                        settingsMgr.setKeyVerified(API, true)
                    }
                }

                settingsMgr.setFollowGPS(dataMap.getBoolean(WearableSettings.KEY_FOLLOWGPS, false))

                val unitMap = dataMap.getDataMap(WearableSettings.KEY_UNITS)
                val oldUnits = settingsMgr.getUnitString()
                if (unitMap != null) {
                    settingsMgr.setTemperatureUnit(unitMap.getString(WearableSettings.KEY_TEMPUNIT, Units.FAHRENHEIT))
                    settingsMgr.setSpeedUnit(unitMap.getString(WearableSettings.KEY_SPEEDUNIT, Units.MILES_PER_HOUR))
                    settingsMgr.setDistanceUnit(unitMap.getString(WearableSettings.KEY_DISTANCEUNIT, Units.MILES))
                    settingsMgr.setPressureUnit(unitMap.getString(WearableSettings.KEY_PRESSUREUNIT, Units.INHG))
                    settingsMgr.setPrecipitationUnit(unitMap.getString(WearableSettings.KEY_PRECIPITATIONUNIT, Units.INCHES))
                } else {
                    settingsMgr.setDefaultUnits(dataMap.getString(WearableSettings.KEY_TEMPUNIT, Units.FAHRENHEIT))
                }
                val newUnits = settingsMgr.getUnitString()

                if (!ObjectsCompat.equals(oldUnits, newUnits)) {
                    localBroadcastMgr.sendBroadcast(Intent(CommonActions.ACTION_SETTINGS_UPDATEUNIT))
                }

                LocaleUtils.setLocaleCode(dataMap.getString(WearableSettings.KEY_LANGUAGE, ""))

                val oldIcons = settingsMgr.getIconsProvider()
                settingsMgr.setIconsProvider(
                    dataMap.getString(
                        WearableSettings.KEY_ICONPROVIDER,
                        WeatherIconsEFProvider.KEY
                    )
                )
                val newIcons = settingsMgr.getIconsProvider()
                if (!ObjectsCompat.equals(oldIcons, newIcons)) {
                    // Update tiles and complications
                    requestWidgetUpdate(appContext)
                }

                val devSettingsMap = dataMap.getDataMap(WearableSettings.KEY_DEVSETTINGS)
                if (devSettingsMap != null) {
                    DevSettingsEnabler.clearPreferences(
                        appContext,
                        devSettingsMap.getBoolean(WearableSettings.KEY_DEVSETTINGS, false)
                    )
                } else {
                    DevSettingsEnabler.clearPreferences(appContext, false)
                }

                val apiKeyMap = dataMap.getDataMap(WearableSettings.KEY_APIKEYS)
                val apiKeyVerifyMap = dataMap.getDataMap(WearableSettings.KEY_APIKEYS_VERIFIED)
                if (apiKeyMap != null) {
                    for (key in apiKeyMap.keySet()) {
                        settingsMgr.setAPIKey(
                            key,
                            apiKeyMap.getString(key)
                        )
                        if (apiKeyVerifyMap != null) {
                            settingsMgr.setKeyVerified(key, apiKeyVerifyMap.getBoolean(key, false))
                        }
                    }
                }

                setSettingsUpdateTime(appContext, updateTimeMillis)

                Timber.tag(TAG).d("Updated settings")
            }

            // Send callback to receiver
            localBroadcastMgr.sendBroadcast(Intent(WearableHelper.SettingsPath))
        }
    }

    suspend fun updateLocation(context: Context, dataMap: DataMap?) = withContext(Dispatchers.Default) {
        val appContext = context.applicationContext
        val settingsMgr = SettingsManager(appContext)

        if (!isActive) return@withContext

        if (dataMap != null && !dataMap.isEmpty) {
            val locationJSON = dataMap.getString(WearableSettings.KEY_LOCATIONDATA, "")
            if (!locationJSON.isNullOrBlank()) {
                val locationData = JSONParser.deserializer(locationJSON, LocationData::class.java)

                if (locationData != null) {
                    val updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME)

                    if (updateTimeMillis != getLocationDataUpdateTime(appContext) ||
                            locationData != settingsMgr.getHomeData()) {
                        settingsMgr.updateLocation(locationData)
                    }

                    setLocationDataUpdateTime(appContext, updateTimeMillis)

                    Timber.tag(TAG).d("updateLocation: Updated location data")

                    // Send callback to receiver
                    LocalBroadcastManager.getInstance(appContext).sendBroadcast(
                            Intent(WearableHelper.LocationPath))
                }
            }
        }
    }

    suspend fun updateWeather(context: Context, dataMap: DataMap?) = withContext(Dispatchers.Default) {
        val appContext = context.applicationContext
        val settingsMgr = SettingsManager(appContext)
        val weatherDAO = WeatherDatabase.getWeatherDAO(appContext)

        if (!isActive) return@withContext

        if (dataMap != null && !dataMap.isEmpty) {
            val updateTimeMillis = dataMap.getLong(WearableSettings.KEY_UPDATETIME)
            // Check if data actually exists to force an update
            var dataExists = false
            val homeData = settingsMgr.getHomeData()
            if (homeData != null) {
                dataExists = withContext(Dispatchers.IO) {
                    weatherDAO.getWeatherDataCountByKey(homeData.query) > 0
                }
            }
            if (updateTimeMillis != getWeatherUpdateTime(appContext) || !dataExists) {
                val weatherAsset = dataMap.getAsset(WearableSettings.KEY_WEATHERDATA)
                if (weatherAsset != null) {
                    try {
                        withContext(Dispatchers.IO) {
                            val fd = Wearable.getDataClient(appContext).getFdForAsset(weatherAsset).await()
                            fd.inputStream.use { inputStream ->
                                val weatherData = JSONParser.deserializer(inputStream, Weather::class.java)
                                if (weatherData?.isValid == true) {
                                    settingsMgr.saveWeatherAlerts(homeData, weatherData.weatherAlerts)
                                    settingsMgr.saveWeatherData(weatherData)
                                    settingsMgr.saveWeatherForecasts(Forecasts(weatherData))
                                    settingsMgr.saveWeatherForecasts(weatherData.query, weatherData.hrForecast?.map { input -> HourlyForecasts(weatherData.query, input) })
                                    settingsMgr.setUpdateTime(LocalDateTime.ofInstant(Instant.ofEpochMilli(updateTimeMillis), ZoneOffset.UTC))
                                    setWeatherUpdateTime(appContext, updateTimeMillis)

                                    Timber.tag(TAG).d("Updated weather data")

                                    requestWidgetUpdate(appContext)
                                } else {
                                    Timber.tag(TAG).d("Weather data invalid")
                                }
                            }
                        }
                    } catch (e: IOException) {
                        Timber.e(e)
                    } catch (e: Exception) {
                        if (e is ApiException || e.cause is ApiException) {
                            val apiException = e.cause as? ApiException ?: e as? ApiException
                            if (apiException?.statusCode != WearableStatusCodes.API_NOT_CONNECTED &&
                                apiException?.statusCode != WearableStatusCodes.TARGET_NODE_NOT_CONNECTED
                            ) {
                                Timber.e(e)
                            }
                        }
                    }
                } else {
                    Timber.tag(TAG).d("updateWeather: weather data missing")
                }
            }

            // Send callback to receiver
            LocalBroadcastManager.getInstance(appContext).sendBroadcast(
                Intent(WearableHelper.WeatherPath).apply {
                    putExtra(
                        WearableSettings.KEY_PARTIAL_WEATHER_UPDATE,
                        dataMap.getBoolean(WearableSettings.KEY_PARTIAL_WEATHER_UPDATE)
                    )
                }
            )
        }
    }

    private fun getSettingsUpdateTime(context: Context): Long {
        val prefs = context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        return prefs.getLong("settings_updatetime", 0)
    }

    private fun getLocationDataUpdateTime(context: Context): Long {
        val prefs = context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        return prefs.getLong("location_updatetime", 0)
    }

    private fun getWeatherUpdateTime(context: Context): Long {
        val prefs = context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        return prefs.getLong("weather_updatetime", 0)
    }

    private fun setSettingsUpdateTime(context: Context, value: Long) {
        val prefs =
            context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        prefs.edit {
            putLong("settings_updatetime", value)
        }
    }

    private fun setLocationDataUpdateTime(context: Context, value: Long) {
        val prefs =
            context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        prefs.edit {
            putLong("location_updatetime", value)
        }
    }

    private fun setWeatherUpdateTime(context: Context, value: Long) {
        val prefs =
            context.applicationContext.getSharedPreferences("datasync", Context.MODE_PRIVATE)
        prefs.edit {
            putLong("weather_updatetime", value)
        }
    }
}